platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileCompactor.java

```diff
@@ -0,0 +1,470 @@
+import java.io.IOException;
+import java.nio.channels.ClosedByInterruptException;
+import java.nio.file.Path;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * This class is responsible performing compaction of data files in a {@link DataFileCollection}.
+ * The compaction is supposed to happen in the background and can be paused and resumed with {@link #pauseCompaction()}
+ * and {@link #resumeCompaction()} to prevent compaction from interfering with snapshots.
+ */
+public class DataFileCompactor {
+
+    private static final Logger logger = LogManager.getLogger(DataFileCompactor.class);
+
+    /**
+     * Since {@code com.swirlds.platform.Browser} populates settings, and it is loaded before any
+     * application classes that might instantiate a data source, the {@link ConfigurationHolder}
+     * holder will have been configured by the time this static initializer runs.
+     */
+    private static final MerkleDbConfig config = ConfigurationHolder.getConfigData(MerkleDbConfig.class);
+
+    public static final int DEFAULT_COMPACTION_LEVEL = 0;
+    /**
+     * The data file collection to compact
+     */
+    private final DataFileCollection<?> dataFileCollection;
+
+    /**
+     * A lock used for synchronization between snapshots and compactions. While a compaction is in
+     * progress, it runs on its own without any synchronization. However, a few critical sections
+     * are protected with this lock: to create a new compaction writer/reader when compaction is
+     * started, to copy data items to the current writer and update the corresponding index item,
+     * and to close the compaction writer. This mechanism allows snapshots to effectively put
+     * compaction on hold, which is critical as snapshots should be as fast as possible, while
+     * compactions are just background processes.
+     */
+    private final Semaphore snapshotCompactionLock = new Semaphore(1);
+
+    /**
+     * Start time of the current compaction, or null if compaction isn't running
+     */
+    private final AtomicReference<Instant> currentCompactionStartTime = new AtomicReference<>();
+    /**
+     * Current data file writer during compaction, or null if compaction isn't running. The writer
+     * is created at compaction start. If compaction is interrupted by a snapshot, the writer is
+     * closed before the snapshot, and then a new writer / new file is created after the snapshot is
+     * taken.
+     */
+    private final AtomicReference<DataFileWriter<?>> currentCompactionWriter = new AtomicReference<>();
```

Comment:
`currentCompactionWriter` did make sense in context of `DataFileCollection` to distinguish it from the flush writer. Now in this class it can be named just `currentWriter`, similar for `currentReader`.

Reply:
Agreed. I renamed it